src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+        private void StartGetModelTask()
+        {
+            _getModelQueue.AddWork();
+        }
+
+        /// <summary>
+        /// Starts a new task to update the current document model.
+        /// </summary>
+        private void StartModelUpdateTask()
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+                _updateModelQueue.AddWork();
+        }
+
+        /// <summary>
+        /// Starts a new task to highlight the symbol node corresponding to the current caret position in the editor.
+        /// </summary>
+        private void StartHightlightNodeTask()
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+                _highlightNodeQueue.AddWork();
+        }
+
+        private void ExpandAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource, true);
+        }
+
+        private void CollapseAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource, false);
+        }
+
+        private void Search(object sender, EventArgs e)
+        {
+            StartModelUpdateTask();
+        }
+
+        private void SortByName(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Name;
+            StartModelUpdateTask();
+        }
+
+        private void SortByOrder(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Order;
+            StartModelUpdateTask();
+        }
+
+        private void SortByType(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Type;
+            StartModelUpdateTask();
+        }
+
+        // When symbol node clicked, select the corresponding code
+        private void JumpToContent(object sender, EventArgs e)
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+            {
+                CodeWindow.GetLastActiveView(out var textView);
+                var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+                if (activeTextView is not null && sender is StackPanel panel && panel.DataContext is DocumentSymbolViewModel symbol)
+                {
+                    // Avoids highlighting the node after moving the caret ourselves 
+                    // (The node is already highlighted on user click)
+                    activeTextView.Caret.PositionChanged -= Caret_PositionChanged;
+
+                    // Get the position of the start of the line the symbol is on
+                    var position = LspSnapshot.GetLineFromLineNumber(symbol.StartPosition.Line).Start.Position;
+
+                    // Gets a point for this position with respect to the updated snapshot
+                    var currentSnapshot = activeTextView.TextSnapshot;
+                    var snapshotPoint = new SnapshotPoint(currentSnapshot, position);
```

Comment:
Shouldn't this be doing some sort of tracking span or something? Because if I delete a bunch of code, the position could now be off the end of the file or something?